### DISCUSSION TITLE
Pyodide pyfetch giving name error

### DISCUSSION THREAD

**Discussion by N-Prize** [pyodide/pyodide#4117] [General]
```py
from pyodide.http import pyfetch
import pandas as pd

filename = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%204/data/example1.txt"

async def download(url, filename):
    response = await pyfetch(url)
    
    if response.status == 200:
        with open(filename, 'wb') as f:  # 'wb' for write binary mode
            f.write(await response.bytes())

await download(filename, 'example1.txt')

print("done") 
```

Despite many attempts to install pyodide "correctly", this code gives a "NameError: name '_jsfetch' is not defined"
What could be the underlying problem?

**Comment by ryanking13** [comment]
Could you share us the following?

1) The Pyodide version you are using
2) The environment that you are running the Pyodide (Chrome, Firefox, Safari, Node.js, WebWorker, ServiceWorker, etc...)

**Reply by Matinnorouzi2023** [reply]
Hi, I have the same problem too,

`import pyodide
print(pyodide.__version__) `

pyodide version is 0.24.1
I am using the Mozilla browser 


This is more information from my error : 
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[5], line 17
     13         with open(filename, "wb") as f:
     15             f.write(await response.bytes())
---> 17 await download(filename, "example1.txt")
     19 print("done")

Cell In[5], line 9, in download(url, filename)
      7 async def download(url, filename):
----> 9     response = await pyfetch(url)
     11     if response.status == 200:
     13         with open(filename, "wb") as f:

File ~\anaconda3\Lib\site-packages\pyodide\http.py:305, in pyfetch(url, **kwargs)
    274 r"""Fetch the url and return the response.
    275 
    276 This functions provides a similar API to :js:func:`fetch` however it is
   (...)
    301 'version': '0.23.4', 'python': '3.11.2'}, ... # long output truncated
    302 """
    303 try:
    304     return FetchResponse(
--> 305         url, await _jsfetch(url, to_js(kwargs, dict_converter=Object.fromEntries))
    306     )
    307 except JsException as e:
    308     raise OSError(e.message) from None

NameError: name '_jsfetch' is not defined

**Reply by ryanking13** [reply]
> ~\anaconda3\Lib\site-packages\pyodide\http.py

Looking at the stacktrace, it seems like you are running Pyodide in the local machine not in the browser?

**Reply by Matinnorouzi2023** [reply]
Yes, It is on the local host browser

**Reply by ryanking13** [reply]
Could you please tell us how we can reproduce the error?

**Reply by Matinnorouzi2023** [reply]
First time I Faced with this error : 
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[40], line 1
----> 1 import piplite
      2 await piplite.install(['xlrd','openpyxl'])

ModuleNotFoundError: No module named 'piplite'



So, I update anaconda and python and also update pipeline but it isn't compatible.

**Reply by ryanking13** [reply]
`piplite` is a JupyterLite only module. Are you running your code in the jupyterlite environment?

**Reply by Matinnorouzi2023** [reply]
No, I am not using  jupyterlite environment. You are right. oh my god, Thank you so much for your help.

**Reply by Matinnorouzi2023** [reply]
is suitable to use the online JupyterLite environment for this function?

**Reply by ryanking13** [reply]
Jupyterlite uses Pyodide internally, so yes, pyfetch would work there.

**Reply by Matinnorouzi2023** [reply]
Yes, It works smoothly thank you for sharing your information.

**Reply by Matinnorouzi2023** [reply]
How I can solve this error on local Jupyter? Does it mean how I can add Jupterlite on "Jupyter lab" of Anaconda or another local host of Python? or another IDE like Google Colab? due to I cannot change my IDE just for an extra library ??

**Reply by ryanking13** [reply]
> How I can solve this error on local Jupyter?

You can't. Pyodide is not a single Python library but a Python that runs in the browser. Since functions like `pyfetch` uses browser Web APIs, it is not available in local Python. Note that `pyfetch` is just a workaround to send HTTP requests in the browser since socket is not available in the browser. If you are running a local host Python, you can use libraries like `urllib`, `requests`.

By the way, if you have a local Python like Jupyterlab, Google Colab, then I think you don't need Pyodide at all as you can run Python codes with local Python not in Pyodide.

**Reply by Matinnorouzi2023** [reply]
It is interesting. Thank you for sharing your information 😊

**Reply by ilartstu** [reply]
I have a similar code and I rewrite it this way with library requests (maybe it will be useful for future people):

```
import requests
import pandas as pd

# URL of file
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/addresses.csv"

# Download file
response = requests.get(url)
filename = "addresses.csv"

# Save file
with open(filename, "wb") as f:
    f.write(response.content)

# Read file by pandas
df = pd.read_csv(filename, header=None)

# First a few strings
print(df.head())
```

**Reply by AbdelrahmanEldoma** [reply]
> I have a similar code and I rewrite it this way with library requests (maybe it will be useful for future people):
> 
> ```
> import requests
> import pandas as pd
> 
> # URL of file
> url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/addresses.csv"
> 
> # Download file
> response = requests.get(url)
> filename = "addresses.csv"
> 
> # Save file
> with open(filename, "wb") as f:
>     f.write(response.content)
> 
> # Read file by pandas
> df = pd.read_csv(filename, header=None)
> 
> # First a few strings
> print(df.head())
> ```

it was useful for me thank you @cho-nado

**Reply by ngonied** [reply]
Thank you @cho-nado

**Reply by CarlRP** [reply]
@cho-nado Thanks, I rewrote the original code and with your idea and solution, it looks like this:

```
import requests as rq
import pandas as pd
filename = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/addresses.csv"

def download(url, filename):
    response = rq.get(url)
    if response == 200:
        with open(filename, "wb") as f:
            f.write(response.content)

download(filename, "addresses.csv")

df = pd.read_csv(filename, header=None)

print(df)
```
